### ISSUE TITLE
an operation on two integers would have overflowed and corrupted data

### ISSUE THREAD

**Issue by Niedzwiedzw** [Ryan-rsm-McKenzie/bsa-rs#11]
https://github.com/Niedzwiedzw/hoolamike/blob/1a80ccf0b369ede56edff3614ffb95cfae552361/crates/hoolamike/src/install_modlist/directives/create_bsa.rs#L56

this is where it happens, so pretty much at write

full stacktrace:

```rust
Caused by:
    could not finish installation due to errors:
    handling directive: [Bsa(
        CreateBSADirectiveKind {
            hash: "LhfDPwyDFQU=",
            size: 0,
            to: "mods\\Tuxborn - NPC Merge\\NPC Appearances Merged - Textures.bsa",
            temp_id: "25bff317-7305-4c37-a9ed-c786b779d1d4",
            file_states: [
                Wit]:
    0. handling directive: [Bsa(
        CreateBSADirectiveKind {
            hash: "LhfDPwyDFQU=",
            size: 0,
            to: "mods\\Tuxborn - NPC Merge\\NPC Appearances Merged - Textures.bsa",
            temp_id: "25bff317-7305-4c37-a9ed-c786b779d1d4",
            file_states: [
                Wit]
    1. creating BA2 (fallout4/starfield) archive
    2. writing bsa file to "./mods/Tuxborn - NPC Merge/NPC Appearances Merged - Textures.bsa"
    3. an operation on two integers would have overflowed and corrupted data

```

**Comment by Ryan-rsm-McKenzie** [comment]
I made the branch `overflow-verbose` to help you narrow down exactly where it is failing. Typically this error indicates your archive is too large to be written without loss of information due to integral overflow.

**Comment by Niedzwiedzw** [comment]
```
Caused by:
    could not finish installation due to errors:
    handling directive: [Bsa(
        CreateBSADirectiveKind {
            hash: "LhfDPwyDFQU=",
            size: 0,
            to: "mods\\Tuxborn - NPC Merge\\NPC Appearances Merged - Textures.bsa",
            temp_id: "25bff317-7305-4c37-a9ed-c786b779d1d4",
            file_states: [
                Wit]:
    0. handling directive: [Bsa(
        CreateBSADirectiveKind {
            hash: "LhfDPwyDFQU=",
            size: 0,
            to: "mods\\Tuxborn - NPC Merge\\NPC Appearances Merged - Textures.bsa",
            temp_id: "25bff317-7305-4c37-a9ed-c786b779d1d4",
            file_states: [
                Wit]
    1. creating BSA (skyrim and before) archive
    2. writing bsa file (skyrim and before) to "./mods/Tuxborn - NPC Merge/NPC Appearances Merged - Textures.bsa"
    3. an operation on two integers would have overflowed and corrupted data (overflow on the file data offset ptr while appending the file data)
    ```

**Comment by Niedzwiedzw** [comment]
it seems like you're adding the masked size to the file_data_offset

**Comment by Niedzwiedzw** [comment]
couldn't you use the size that .write() method returns?

**Comment by Ryan-rsm-McKenzie** [comment]
The masked size is the file data size with the file info bits masked off, i.e. the raw size. The archive is too large to be written to the disk without corruption. Integer overflow is annoying to catch in most languages, so this is the only library to catch this error.

**Comment by Niedzwiedzw** [comment]
you probably would not accept a PR which adds a feature flag to ignore the overflows?

**Comment by Niedzwiedzw** [comment]
so I did some digging and this archive output size should be 2094458492

```python
>>> 2094458492 / 4_294_967_295
0.487654118912214
```

**Comment by Niedzwiedzw** [comment]
so my thinking is some of the things I'm putting in there might be too big. so yeah, I'll close this issue since this is definitely not caused by bsa-rs. thank you

**Comment by Ryan-rsm-McKenzie** [comment]
I saw you're working with dds compression. [DirectXTex](https://github.com/microsoft/DirectXTex) supports [conversion](https://github.com/microsoft/DirectXTex/wiki/Convert)/[compression](https://github.com/microsoft/DirectXTex/wiki/Compress)/[decompression](https://github.com/microsoft/DirectXTex/wiki/Decompress) with pretty much every format ever used for dds files, and its written by Chuck Walbourn who's basically _the_ DirectX guy. I wrote [bindings](https://lib.rs/crates/directxtex) to use it in rust since I use it in this library for Fallout 4's DX10 archive format. You may want to consider using that instead of what you're using right now.

**Comment by Niedzwiedzw** [comment]
<pikachu-face.jpg>

**Comment by Niedzwiedzw** [comment]
Soo I thought that the cause for overflow was the fact that all the recompessed/resized textures were 20 bytes too big, but now that I'm using your bindings to  directxtex (which fixes the sizes) the issue still persists 
https://github.com/Niedzwiedzw/hoolamike/blob/feature/directxtex-integration/crates/hoolamike/src/install_modlist/directives/create_bsa/tes_4.rs

this only happens for this one skyrim archive, I have double checked with the author and if it's packed using wabbajack it works fine and all files are present (which means it's not corrupted I guess)

it's no secret that at this point it's pretty much you writing this project for me, but I don't have enough knowledge to debug this I'm afraid. Some modlists work and in general this tool is useful, but it'd be sad if some modlists didn't work.

so long story short, could you take a look at this please?

**Comment by Niedzwiedzw** [comment]
false alarm, I was loading the files wrong :D
